### ISSUE TITLE
Auto-split the Bashed Patch once it has 255 masters

### ISSUE THREAD

**Issue by Infernio** [wrye-bash/wrye-bash#533]
Bethesda's plugin format has a hard limit of 255 masters, since FormIDs only have a single byte for master indices (and 2^8 - 1 = 255).
Right now, we show an error message and tell the user of a workaround, which involves:
1. Creating a new BP after the main one
1. Unchecking patchers until the BP process completes
1. Building the second BP with the patchers you unchecked in the second step
1. If that fails as well, keep repeating from step 1 with the newly split patch

It would be convenient if we could automate this. So when we go to save, do something like this:
1. Ask for all masters used in the patch
1. If those amount to <=255/253 (always the case except for *really* heavy SSE/FO4 load orders), just save it
1. Otherwise, ask each top GRUP in the patch file for its master count
1. If any of those GRUPs have >255/253 masters, error out entirely - splitting GRUPs into multiple sets of records is out of scope for this issue
1. Sort by number of masters and move out the GRUP with the largest number of masters (or maybe we could calculate unique masters?) into a second PatchFile
1. Check if that got us to <=255/253 masters in the first file - if not, repeat the previous step
1. If the first file is now empty, error out entirely as well - clearly there is a more complex relationship going on in such a case than what this simple algorithm can handle
1. Once the first file is done, move on to the second file and repeat from step 1

Not exactly high priority since it's fairly rare and a workaround exists, but would be nice to have.
We should also mark the second BP somehow and prevent rebuilding it - maybe don't add the `BASHED PATCH` author so that it doesn't get recognized as a BP?

**Comment by lojack5** [comment]
Since you started working on this, a few UX things that would be nice to have with this, basically all revolving around the split parts behaving as if all the same bashed patch by default (with some way to "unlink" the parts):
- Rebuilding on any part rebuilds the whole patch
- On rebuild if fewer parts are needed, we should cleanup the leftover parts from last build
- (Maybe?) moving load order of one of the parts should move them as a block of plugins

I'm sure there's more.

**Comment by Infernio** [comment]
Moving them together in LO sounds really tough. But I would like to do some basic UX improvements for this, at *least* the deleting of unneeded parts.

**Comment by selsie** [comment]
This function doesn't work. I got a few error messages, and that's it.
The easiest solution for me is still to have a separate patch for levelled lists.

**Comment by sibir-ine** [comment]
Please post the `BashBugDump.log` in Wrye Bash's installation folder, the devs need to know the actual error messages.

**Comment by selsie** [comment]
localize.py  113 setup_locale: Wrye Bash does not support the language family 'en', will however try to set locale to 'en'
localize.py  125 setup_locale: Set wxPython locale to 'en'
loot_parser.py   54 <module>: Using LibYAML-based parser
doc_viewer.py   57 <module>: pdfviewer using PyMuPDF (GPL)
balt.py   57 <module>: wx.html2.WebView is missing, features utilizing HTML will be disabled
bash.py  327 dump_environment: Using Wrye Bash Version 312.1 (Standalone)
	OS info: Windows-10-10.0.19045, running on Intel64 Family 6 Model 158 Stepping 13, GenuineIntel
	Python version: 3.11.7 (tags/v3.11.7:fa7a6f2, Dec  4 2023, 19:24:49) [MSC v.1937 64 bit (AMD64)]
	Dependency versions:
	 - chardet: 5.2.0
	 - ifileoperation: 1.2.6
	 - lxml: 5.0.0
	 - packaging: 23.2
	 - PyMuPDF: 1.23.8; bundled MuPDF version: 1.23.7
	 - python-lz4: 4.3.3; bundled LZ4 version: 1.9.4
	 - PyYAML: 6.0.1
	 - requests: 2.31.0
	 - vdf: 3.4
	 - websocket-client: 1.7.0
	 - wxPython: 4.2.1 msw (phoenix) wxWidgets 3.2.2.1
	Input encoding: None; output encoding: utf-8
	Filesystem encoding: utf-8
	Command line: ['D:\\Skyrim Tools\\Mopy\\Wrye Bash.exe', '-o', 'D:\\Games\\steamapps\\common\\Skyrim Special Edition', '-L', 'English']
bash.py  610 _import_bush_and_set_game: Searching for game to manage:
bush.py  178 _supportedGames: The following games are supported by this version of Wrye Bash:
  - Enderal (Steam)
  - Enderal Special Edition (GOG, Steam)
  - Fallout 3 (EGS, GOG, Steam, WS)
  - Fallout 4 (GOG, Steam, WS)
  - Fallout 4 VR (Steam)
  - Fallout New Vegas (EGS, GOG, Steam, WS)
  - Morrowind (GOG, Steam, WS)
  - Nehrim (GOG, Steam)
  - Oblivion (GOG, Steam, WS)
  - Skyrim (Steam)
  - Skyrim Special Edition (EGS, GOG, Steam, WS)
  - Skyrim VR (Steam)
  - Starfield (Steam, WS)
Wrye Bash looked for installations of supported games in the following places:
 1. Steam:
  The following supported games were found via Steam:
   - Morrowind (Steam): D:\Games\steamapps\common\Morrowind
   - Oblivion (Steam): D:\Games\steamapps\common\Oblivion
   - Skyrim Special Edition (Steam): D:\Games\steamapps\common\Skyrim Special Edition
 2. GOG (via Windows Registry):
  No supported games were found via GOG.
 3. Windows Store (Legacy):
  No supported games with modding enabled were found via the legacy Windows Store.
 4. Windows Store:
  No supported games were found via the Windows Store.
 5. Epic Games Store:
  No supported games were found via the Epic Games Store.
bush.py  246 _detectGames: Detecting games via the -o argument, bash.ini and relative path:
bush.py  252 _detectGames: Set game mode to Skyrim Special Edition (Steam) specified via -o argument:  D:\Games\steamapps\common\Skyrim Special Edition
bush.py  268 __setGame: Using Skyrim Special Edition (Steam) game: D:\Games\steamapps\common\Skyrim Special Edition
patch_game.py  304 _import_records: Signatures [b'CLDC', b'HAIR', b'PARW', b'PBAR', b'PBEA', b'PCON', b'PFLA', b'PGRE', b'PHZD', b'PMIS', b'PWAT', b'RGDL', b'SCOL', b'SCPT'] lack an implementation - defaulting to MreRecord
initialization.py  141 init_dirs: My Games location set to C:\Users\*****\Dokumente\My Games\Skyrim Special Edition
initialization.py  165 init_dirs: LocalAppData location set to C:\Users\*****\AppData\Local\Skyrim Special Edition
initialization.py  223 init_dirs: Game Mods location set to D:\Games\steamapps\common\Skyrim Special Edition Mods
initialization.py  231 init_dirs: Bash Mod Data location set to D:\Games\steamapps\common\Skyrim Special Edition Mods\Bash Mod Data
initialization.py  238 init_dirs: Installers location set to D:\Games\steamapps\common\Skyrim Special Edition Mods\Bash Installers
initialization.py  241 init_dirs: Installers bash data location set to D:\Skyrim Tools\Mopy\Bash Installers
initialization.py  249 init_dirs: Checking if WB directories exist and creating them if needed:
initialization.py  253 init_dirs:  - D:\Games\steamapps\common\Skyrim Special Edition Mods\Bash Mod Data
initialization.py  253 init_dirs:  - D:\Games\steamapps\common\Skyrim Special Edition Mods\Bash Installers
initialization.py  253 init_dirs:  - D:\Games\steamapps\common\Skyrim Special Edition Mods\Bash Installers\Bain Converters
initialization.py  253 init_dirs:  - D:\Games\steamapps\common\Skyrim Special Edition Mods\Bash Installers\Bain Converters\--Duplicates
initialization.py  253 init_dirs:  - D:\Games\steamapps\common\Skyrim Special Edition Mods\Bash Installers\Bain Converters\--Corrupt
initialization.py  253 init_dirs:  - D:\Skyrim Tools\Mopy\Bash Installers
initialization.py  253 init_dirs:  - D:\Skyrim Tools\Mopy\Bash Installers\BSA Cache
loot_parser.py   87 __init__: Using these LOOT paths:
loot_parser.py   88 __init__:  Masterlist: C:\Users\*****\AppData\Local\LOOT\games\Skyrim Special Edition\masterlist.yaml
loot_parser.py   89 __init__:  Userlist: C:\Users\*****\AppData\Local\LOOT\games\Skyrim Special Edition\userlist.yaml
loot_parser.py   90 __init__:  Taglist (fallback): D:\Skyrim Tools\Mopy\taglists\SkyrimSE\taglist.yaml
loot_parser.py   97 __init__: Initialized loot_parser, compatible with LOOT metadata v0.21
__init__.py 3790 initBosh: Looking for main game INI at C:\Users\*****\Dokumente\My Games\Skyrim Special Edition\Skyrim.ini
windows.py 1158 testUAC: Testing if game folder is UAC-protected
__init__.py 1612 _initDB: Initializing BSAInfos
__init__.py 1613 _initDB:  store_dir: D:\Games\steamapps\common\Skyrim Special Edition\Data
__init__.py 1614 _initDB:  bash_dir: D:\Games\steamapps\common\Skyrim Special Edition Mods\Bash Mod Data\BSA Data
__init__.py 1612 _initDB: Initializing ModInfos
__init__.py 1613 _initDB:  store_dir: D:\Games\steamapps\common\Skyrim Special Edition\Data
__init__.py 1614 _initDB:  bash_dir: D:\Games\steamapps\common\Skyrim Special Edition Mods\Bash Mod Data
_games_lo.py  645 print_lo_paths: Using the following load order files:
_games_lo.py  647 print_lo_paths:  - Load order and active plugins: C:\Users\*****\AppData\Local\Skyrim Special Edition\plugins.txt
__init__.py 1612 _initDB: Initializing SaveInfos
__init__.py 1613 _initDB:  store_dir: C:\Users\*****\Dokumente\My Games\Skyrim Special Edition\Saves
__init__.py 1614 _initDB:  bash_dir: C:\Users\*****\Dokumente\My Games\Skyrim Special Edition\Saves\Bash
__init__.py 1612 _initDB: Initializing INIInfos
__init__.py 1613 _initDB:  store_dir: D:\Games\steamapps\common\Skyrim Special Edition\Data\INI Tweaks
__init__.py 1614 _initDB:  bash_dir: D:\Games\steamapps\common\Skyrim Special Edition Mods\Bash Mod Data\INI Data
__init__.py 3809 __init__: Constructing panel 'Installers'
__init__.py 3818 __init__: Panel 'Installers' constructed successfully
__init__.py 3809 __init__: Constructing panel 'Mods'
__init__.py 3818 __init__: Panel 'Mods' constructed successfully
__init__.py 3809 __init__: Constructing panel 'Saves'
__init__.py 3818 __init__: Panel 'Saves' constructed successfully
__init__.py 3809 __init__: Constructing panel 'INI Edits'
__init__.py 3818 __init__: Panel 'INI Edits' constructed successfully
__init__.py 3809 __init__: Constructing panel 'Screenshots'
__init__.py 1612 _initDB: Initializing ScreenInfos
__init__.py 1613 _initDB:  store_dir: D:\Games\steamapps\common\Skyrim Screenshots
__init__.py 1614 _initDB:  bash_dir: D:\Games\steamapps\common\Skyrim Special Edition Mods\Bash Mod Data\Screenshot Data
__init__.py 3818 __init__: Panel 'Screenshots' constructed successfully
common_records.py  507 mergeWith: Merging changes from mod 'Technicolor Alchemy - CACO - Tundra Berries - More Growable Plants.esp' to leveled list <CACO_ALLIngredients_New [LVLI:(Complete Alchemy & Cooking Overhaul.esp, 6B0F3A)]> caused it to exceed 255 entries. Truncating back to 255, you will have to fix this manually!
patcher_dialog.py  320 _error: Exception during Bashed Patch building:
Traceback (most recent call last):
  File "bash\bolt.py", line 575, in __eq__
AttributeError: 'PatchFile' object has no attribute '_lower'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "bash\basher\patcher_dialog.py", line 215, in PatchExecute
  File "bash\patcher\patch_files.py", line 501, in split_patch
  File "bash\patcher\patch_files.py", line 485, in new_bp_part
  File "bash\bosh\__init__.py", line 3129, in create_new_mod
  File "bash\bosh\__init__.py", line 2305, in cached_lo_insert_after
  File "bash\bolt.py", line 578, in __eq__
TypeError: descriptor 'lower' for 'str' objects doesn't apply to a 'PatchFile' object

**Comment by selsie** [comment]
BTW, the output file Bashed Patch, 0-1.esp is 92BYTES in size.... Content?
`TES4D               ,   HEDR HáÚ?       CNAM
 BASHED PATCH MAST Skyrim.esm DATA     `

**Comment by Infernio** [comment]
I'll take a look once I have some time (aka the weekend), unless someone else gets to this.

**Comment by Utumno** [comment]
Found it will post asap
